File: NutriSmart/app/src/main/java/com/example/nutrismart/data/SessionManager.kt
package com.example.nutrismart.data

import android.content.Context
import android.content.SharedPreferences
import androidx.core.content.edit

class SessionManager(context: Context) {
    private val prefs: SharedPreferences = context.getSharedPreferences("nutrismart_prefs", Context.MODE_PRIVATE)

    fun saveAuthToken(token: String) {
        prefs.edit { putString("USER_TOKEN", token) }
    }

    fun fetchAuthToken(): String? {
        return prefs.getString("USER_TOKEN", null)
    }

    fun saveUserId(userId: Long) {
        prefs.edit { putLong("USER_ID", userId) }
    }

    fun fetchUserId(): Long {
        return prefs.getLong("USER_ID", -1L)
    }

    fun clearSession() {
        prefs.edit { clear() }
    }

    fun saveFastingState(isFasting: Boolean, startTime: Long, endTime: Long, durationHours: Int) {
        prefs.edit {
            putBoolean("FASTING_IS_ACTIVE", isFasting)
                .putLong("FASTING_START_TIME", startTime)
                .putLong("FASTING_END_TIME", endTime)
                .putInt("FASTING_DURATION_HOURS", durationHours)
        }
    }

    fun isFastingActive(): Boolean = prefs.getBoolean("FASTING_IS_ACTIVE", false)
    fun getFastingStartTime(): Long = prefs.getLong("FASTING_START_TIME", 0L)
    fun getFastingEndTime(): Long = prefs.getLong("FASTING_END_TIME", 0L)
    fun getFastingDurationHours(): Int = prefs.getInt("FASTING_DURATION_HOURS", 16)

    fun clearFastingState() {
        prefs.edit {
            remove("FASTING_IS_ACTIVE")
                .remove("FASTING_START_TIME")
                .remove("FASTING_END_TIME")
        }
    }

    fun saveWaterIntake(ml: Int) {
        val today = java.time.LocalDate.now().toString()
        prefs.edit {
            putInt("WATER_CONSUMED", ml)
            putString("WATER_DATE", today)
        }
    }

    fun getWaterIntake(): Int {
        val savedDate = prefs.getString("WATER_DATE", "")
        val today = java.time.LocalDate.now().toString()
        return if (savedDate == today) {
            prefs.getInt("WATER_CONSUMED", 0)
        } else {
            0
        }
    }

    fun saveWakeUpTime(time: String) {
        prefs.edit { putString("wake_up_time", time) }
    }

    fun getWakeUpTime(): String {
        return prefs.getString("wake_up_time", "08:00") ?: "08:00"
    }

    fun saveProfileComplete(isComplete: Boolean) {
        prefs.edit { putBoolean("IS_PROFILE_COMPLETE", isComplete) }
    }

    fun isProfileComplete(): Boolean {
        return prefs.getBoolean("IS_PROFILE_COMPLETE", false)
    }

    fun saveIsVerified(isVerified: Boolean) {
        prefs.edit { putBoolean("IS_VERIFIED", isVerified) }
    }

    fun isVerified(): Boolean {
        return prefs.getBoolean("IS_VERIFIED", false)
    }

    fun saveIsGoogleUser(isGoogle: Boolean) {
        prefs.edit { putBoolean("IS_GOOGLE_USER", isGoogle) }
    }

    fun isGoogleUser(): Boolean {
        return prefs.getBoolean("IS_GOOGLE_USER", false)
    }
}

File: NutriSmartIOS/data/SessionManager.swift
//
//  SessionManager.swift
//  NutriSmartIOS
//
//  Created by Alex on 25/05/2026.
//

import Foundation

class SessionManager {
    static let shared = SessionManager()
    
    private let defaults = UserDefaults.standard
    
    private init() {}

    private let keyUserToken = "USER_TOKEN"
    private let keyUserId = "USER_ID"
    private let keyFastingIsActive = "FASTING_IS_ACTIVE"
    private let keyFastingStartTime = "FASTING_START_TIME"
    private let keyFastingEndTime = "FASTING_END_TIME"
    private let keyFastingDurationHours = "FASTING_DURATION_HOURS"
    private let keyWaterConsumed = "WATER_CONSUMED"
    private let keyWaterDate = "WATER_DATE"
    private let keyWakeUpTime = "wake_up_time"
    private let keyIsProfileComplete = "IS_PROFILE_COMPLETE"
    private let keyIsVerified = "IS_VERIFIED"
    private let keyIsGoogleUser = "IS_GOOGLE_USER"
    
    func saveAuthToken(_ token: String) {
        defaults.set(token, forKey: keyUserToken)
    }
    
    func fetchAuthToken() -> String? {
        return defaults.string(forKey: keyUserToken)
    }
    
    func saveUserId(_ userId: Int) {
        defaults.set(userId, forKey: keyUserId)
    }
    
    func fetchUserId() -> Int {
        return defaults.object(forKey: keyUserId) as? Int ?? -1
    }
    
    func clearSession() {
        if let bundleID = Bundle.main.bundleIdentifier {
            defaults.removePersistentDomain(forName: bundleID)
        }
    }
    
    func saveFastingState(isFasting: Bool, startTime: Int, endTime: Int, durationHours: Int) {
        defaults.set(isFasting, forKey: keyFastingIsActive)
        defaults.set(startTime, forKey: keyFastingStartTime)
        defaults.set(endTime, forKey: keyFastingEndTime)
        defaults.set(durationHours, forKey: keyFastingDurationHours)
    }
    
    func isFastingActive() -> Bool {
        return defaults.bool(forKey: keyFastingIsActive)
    }
    
    func getFastingStartTime() -> Int {
        return defaults.integer(forKey: keyFastingStartTime)
    }
    
    func getFastingEndTime() -> Int {
        return defaults.integer(forKey: keyFastingEndTime)
    }
    
    func getFastingDurationHours() -> Int {
        let duration = defaults.integer(forKey: keyFastingDurationHours)
        return duration == 0 ? 16 : duration
    }
    
    func clearFastingState() {
        defaults.removeObject(forKey: keyFastingIsActive)
        defaults.removeObject(forKey: keyFastingStartTime)
        defaults.removeObject(forKey: keyFastingEndTime)
    }
    
    func saveWaterIntake(ml: Int) {
        let today = getCurrentDateString()
        defaults.set(ml, forKey: keyWaterConsumed)
        defaults.set(today, forKey: keyWaterDate)
    }
    
    func getWaterIntake() -> Int {
        let savedDate = defaults.string(forKey: keyWaterDate) ?? ""
        let today = getCurrentDateString()
        
        if savedDate == today {
            return defaults.integer(forKey: keyWaterConsumed)
        } else {
            return 0
        }
    }

    private func getCurrentDateString() -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter.string(from: Date())
    }
    
    func saveWakeUpTime(_ time: String) {
        defaults.set(time, forKey: keyWakeUpTime)
    }
    
    func getWakeUpTime() -> String {
        return defaults.string(forKey: keyWakeUpTime) ?? "08:00"
    }
    
    func saveProfileComplete(_ isComplete: Bool) {
        defaults.set(isComplete, forKey: keyIsProfileComplete)
    }
    
    func isProfileComplete() -> Bool {
        return defaults.bool(forKey: keyIsProfileComplete)
    }
    
    func saveIsVerified(_ isVerified: Bool) {
        defaults.set(isVerified, forKey: keyIsVerified)
    }
    
    func isVerified() -> Bool {
        return defaults.bool(forKey: keyIsVerified)
    }
    
    func saveIsGoogleUser(_ isGoogle: Bool) {
        defaults.set(isGoogle, forKey: keyIsGoogleUser)
    }
    
    func isGoogleUser() -> Bool {
        return defaults.bool(forKey: keyIsGoogleUser)
    }
}
